packages/zero-client/src/client/connection-manager.ts

```diff
@@ -15,34 +20,242 @@ export type ConnectionState =
+   */
+  #timeoutInterval: ReturnType<typeof setInterval> | undefined;
+
+  /**
+   * Interval duration for checking whether the connecting timeout has elapsed.
+   */
+  #timeoutCheckIntervalMs: number;
+
+  constructor(options: ConnectionManagerOptions) {
+    super();
+
+    const now = Date.now();
+
+    this.#disconnectTimeoutMs = options.disconnectTimeoutMs;
+    this.#timeoutCheckIntervalMs =
+      options.timeoutCheckIntervalMs ?? DEFAULT_TIMEOUT_CHECK_INTERVAL_MS;
+    this.#state = {
+      name: ConnectionStatus.Connecting,
+      attempt: 0,
+      disconnectAt: now + this.#disconnectTimeoutMs,
+    };
+    this.#connectingStartedAt = now;
+    this.#startTimeoutInterval();
+  }
 
   get state(): ConnectionState {
     return this.#state;
   }
 
   /**
-   * Updates the connection status. Returns true if the status changed.
+   * Returns true if the current state is equal to the given status.
+   */
+  is(status: ConnectionStatus): boolean {
+    return this.#state.name === status;
+  }
+
+  /**
+   * Returns true if the run loop should continue.
+   * The run loop continues in disconnected, connecting, and connected states.
+   * It stops in closed state.
+   */
+  shouldContinueRunLoop(): boolean {
+    return this.#state.name !== ConnectionStatus.Closed;
+  }
+
+  /**
+   * Transition to connecting state.
+   *
+   * This starts the 5-minute timeout timer, but if we've entered disconnected state,
+   * we stay there and continue retrying.
    */
-  setStatus(status: ConnectionStatus): boolean {
+  connecting(reason?: string | undefined): void {
     // cannot transition from closed to any other status
     if (this.#state.name === ConnectionStatus.Closed) {
-      return false;
+      return;
     }
 
-    if (status === this.#state.name) {
-      return false;
+    // if we're already in disconnected state, no-op
+    if (this.#state.name === ConnectionStatus.Disconnected) {
+      return;
+    }
+
+    const now = Date.now();
+
+    // If we're already connecting, increment the attempt counter
+    if (this.#state.name === ConnectionStatus.Connecting) {
+      this.#state = {
+        ...this.#state,
+        attempt: this.#state.attempt + 1,
+        reason,
+      };
+      this.notify(this.#state);
+      this.#startTimeoutInterval();
+      return;
+    }
+
+    // Starting a new connecting session
+    // If #connectingStartedAt is undefined, this is a fresh start - set it to now
+    // If it's already set, we're retrying within the same retry window, so keep it
+    if (this.#connectingStartedAt === undefined) {
+      this.#connectingStartedAt = now;
     }
 
-    this.#state = {name: status};
+    const disconnectAt = this.#connectingStartedAt + this.#disconnectTimeoutMs;
+
+    this.#state = {
+      name: ConnectionStatus.Connecting,
+      attempt: 1,
+      disconnectAt,
+      reason,
+    };
     this.notify(this.#state);
-    return true;
+    this.#startTimeoutInterval();
```

Comment:
```suggestion
    this.#maybeStartTimeoutInterval();
```

or


```suggestion
    this.#startTimeoutIntervalIfNeeded();
```

When I read it I was like why are we starting the interval again here?